### ISSUE TITLE
Testing S3 plugin

### ISSUE THREAD

**Issue by hms** [janko-m/shrine#115]
I've written a small enhancement to the S3 plugin to get access to S3.metadata (code below).  I've been unable to make the S3 tests work (I'm using my S3 credentials, so I know I have access to S3).  It gets lost in after { @s3.clear! }

The code has been tested via my application, so I know it works.  But it would be nice to make a working test and a proper pull request.

Hal

```
class Shrine::Storage::S3
 def storage_metadata(id)
        object(id).metadata
      end
end
```

```
class Shrine
  module FileMethods
   def storage_metadata
          raise Error, "#{storage.inspect} does not support the metadata method" unless storage.respond_to?(:metadata)
          storage.metadata(id)
        end
  end
end
```

**Comment by janko-m** [comment]
@hms Could you tell me what exactly fails? I realized now that I forgot to document that you have to have a `.env` file with the following credentials:

```sh
# .env
S3_BUCKET="..."
S3_REGION="..."
S3_ACCESS_KEY_ID="..."
S3_SECRET_ACCESS_KEY="..."
```

Might I ask what is your use case in fetching S3 object metadata?

**Comment by hms** [comment]
@janko-m 

I set up the .env file and have all of the necessary credentials to access s3 (I took them from another working project, so I'm confident that the credentials are valid).  And since the failure is in the S3#clear! method, I'm convinced something else going on that's preventing the S3 tests from working.

My requirement for S3 metadata is I'm tagging the file with extra data via 'x-amz-meta-' headers that allow me to quickly check on state about the file located on S3.  I support both a quick check, which reads the S3::Object.metadata info and a slow check which downloads the file and recomputes the values -- as this is both slow and expensive, so I want to be selective when I do this.

I've added the custom metadata tags on S3 upload (thanks to your earlier help).  But now what I want is to be able to easily have access to them.  It's nice from a code cleanliness standpoint to have the method accessible just like all the other S3 checks (i.e., exists?).

**Comment by janko-m** [comment]
> I set up the .env file and have all of the necessary credentials to access s3 (I took them from another working project, so I'm confident that the credentials are valid). And since the failure is in the S3#clear! method, I'm convinced something else going on that's preventing the S3 tests from working.

If I understood correctly, the tests are hanging, right? There is one slow test where I called `Shrine::Storage::Linter` twice, so it can make it appear as if though the test suite is hanging. Sorry for that, I pushed a fix for that now.

> My requirement for S3 metadata is I'm tagging the file with extra data via 'x-amz-meta-' headers that allow me to quickly check on state about the file located on S3. I support both a quick check, which reads the S3::Object.metadata info and a slow check which downloads the file and recomputes the values -- as this is both slow and expensive, so I want to be selective when I do this.

What is the advantage of reading these metadata values from S3, instead of saving them as Shrine metadata? This way you never have to make any additional HTTP requests to S3. 

> I've added the custom metadata tags on S3 upload (thanks to your earlier help). But now what I want is to be able to easily have access to them. It's nice from a code cleanliness standpoint to have the method accessible just like all the other S3 checks (i.e., exists?).

I considered allowing `#metadata` in the storage abstraction, but for different purposes. I wanted to add it on `Shrine::Storage::Cloudinary`, which extracts actual file metadata on upload (unlike S3), and that would be an optimization for the `restore_cached_data` plugin. However, I since I currently don't see a good use case of fetching S3 metadata, I'm reluctant to have it added to `Shrine::Storage::S3`.

**Comment by hms** [comment]
@janko-m 
The tests aren't hanging -- they are crashing in the s3.clear! method.

I'm not sure there is value in storing my metadata as shrine metadata (vs. just regular database fields) since that's not easily searchable via AR.  I know postgres can search JSON fields, but I'm pretty sure mySQL can't and I'm pretty siure the Amazon Aurora versions of both databases can't too.

The value of adding metadata to the object stored on S3 is that they are part of the same object vs. data associated with the S3 object (e.g., in the database).  In a perfect world, I would have transactional semantics between my application metadata (e.g., the database record that described this object) and the S3 object itself.  Unfortunately, that's possible at the moment.  

So the question for me is: do I want/need data associated with the S3 object (beyond its name) in order to create a greater confidence that the object I'm getting from S3 is the actual object my application thinks it requested?  It turns out that for my application (for reasons I won't go into), the answer is yes.  I do need additional assurances that the source S3 object hasn't been altered or replaced.   The simple solution is to compute a hash of the source object as part of the uploading process.

Now I can easily check that the application is getting what it thinks from S3.  But this is slow and expensive.  So, my 'fast path' is to store the hash as S3 metadata.  This gives me a quicker and a little less reliable check -- but one that's good enough for my typical use case.  And I can still run periodic 'slow path' checks that aren't in the end-user workflow, that verify everything is as expected.

**Comment by janko-m** [comment]
> The tests aren't hanging -- they are crashing in the s3.clear! method.

Ok, then I'll ask you again, what is the error that you're getting? I cannot help you if you won't tell me what actually happened.

> So the question for me is: do I want/need data associated with the S3 object (beyond its name) in order to create a greater confidence that the object I'm getting from S3 is the actual object my application thinks it requested? It turns out that for my application (for reasons I won't go into), the answer is yes.

Since this isn't related to Shrine itself, and it's a very specific use case, I don't think that we should add a new storage method for that. I would recommend that you just use `aws-sdk` directly, using the `Shrine::Storage::S3#object` method that I just made public in the latest commit:

```rb
uploaded_file.storage.object(uploaded_file.id).metadata
```

You can also use the `module_include` plugin to add a `UploadedFile#storage_metadata` if you want to.

**Comment by hms** [comment]
@janko-m 

I recreated my fork of Shrine and was able to run some of the tests.  So it looks like pilot error on my part.  While I did include my existing projects S3 credentials, I forgot they have very restrictive permissions that are appropriate for my application and less than test tests require. On a side note, I wasn't trying to be obtuse about the testing error, I didn't realize you were asking for something like a stack trace.

Having access to the storage.object is definitely good enough. Thanks.
